xmake/rules/vala/xmake.lua

```diff
@@ -103,22 +99,48 @@ rule("vala.build")
                 table.insert(argv, path(headerfile))
             end
         end
-        local vapidir = target:data("vala.vapidir")
+
+        local vapidir = target:values("vala.vapidir")
         if vapidir then
             table.insert(argv, path(vapidir, function (p) return "--vapidir=" .. p end))
         end
-        local valaflags = target:data("vala.flags")
+
+        local valaflags = target:values("vala.flags")
         if valaflags then
             table.join2(argv, valaflags)
         end
-        table.insert(argv, path(sourcefile_vala))
+
+        -- iterating through source files,
+        -- otherwise valac would fail when compiling multiple files
+        for _, sourcefile in ipairs(sourcebatch.sourcefiles) do
+            -- if it's only a vala file
+            if path.extension(sourcefile) == ".vala" then
+                table.insert(argv, path(sourcefile))
+            end
+        end
+
+        batchcmds:show_progress(opt.progress, "${color.build.object}compiling.vala")
         batchcmds:vrunv(valac.program, argv)
-        batchcmds:compile(sourcefile_c, objectfile)
+    end)
+
+    on_buildcmd_file(function (target, batchcmds, sourcebatch, opt)
+        -- Again, only vala files need special treatment
+        if path.extension(sourcebatch) == ".vala" then
```

Comment:
please rename `sourcebatch` to `sourcefile` in on_buildcmd_file

sourcebatch is only for on_buildcmd_files